### PR TITLE
Fix/model upgrade reset init containers

### PR DESCRIPTION
**Pull request by ycliuhw**
### Checklist

 - [ ] ~Checked if it requires a [pylibjuju](https://github.com/juju/python-libjuju) change?~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR?~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed?~
 - [x] Do comments answer the question of why design decisions were made?

----

## Description of change

This PR ensures init containers are upgraded to the current model version when the model is upgraded.

Drive-by fix: no-ops for setStatus for a k8s controller node.

## QA steps

```sh
# bootstrap to 2.7 then deploy k8s workload to a 2.7 model;

$ mkubectl -nt1 get statefulset.apps/mariadb-k8s-operator -o json | jq '.spec.template.spec.containers[].image'
"ycliuhw/jujud-operator:2.7.7"

$ mkubectl -nt1 get statefulset.apps/mariadb-k8s -o json | jq '.spec.template.spec.initContainers[].image'
"ycliuhw/jujud-operator:2.7.7"

$ juju upgrade-controller --agent-stream=develop
best version:
    2.8-rc16
started upgrade to 2.8-rc16

$ juju upgrade-model -m t1 --agent-stream=develop
best version:
    2.8-rc16
started upgrade to 2.8-rc16

$ mkubectl -nt1 get statefulset.apps/mariadb-k8s-operator -o json | jq '.spec.template.spec.containers[].image'
"ycliuhw/jujud-operator:2.8-rc16"

$ mkubectl -nt1 get statefulset.apps/mariadb-k8s -o json | jq '.spec.template.spec.initContainers[].image'
"ycliuhw/jujud-operator:2.8-rc16"

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/1874506

### CODE REVIEW

File: caas/kubernetes/provider/volume.go [hunk 1]
```diff
@@ -430,19 +430,34 @@ func getMountPathForFilesystem(i int, appName string, fs storage.KubernetesFiles
 }
 
 // pushUniqueVolume ensures to only add unique volumes because k8s will not schedule pods if it has duplicated volumes.
```
Comment (babbageclunk) [inline caas/kubernetes/provider/volume.go:432]: Update the comment with the force behaviour?

File: worker/caasunitprovisioner/application_worker.go [hunk 2]
```diff
@@ -98,8 +98,8 @@ func (aw *applicationWorker) loop() error {
 		brokerUnitsWatcher watcher.NotifyWatcher
 		brokerUnitsChannel watcher.NotifyChannel
 
-		appOperatorWatcher  watcher.NotifyWatcher
-		appOpertatorChannel watcher.NotifyChannel
+		appOperatorWatcher watcher.NotifyWatcher
+		appOperatorChannel watcher.NotifyChannel
```
Comment (babbageclunk) [inline worker/caasunitprovisioner/application_worker.go:102]: :+1:

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -1430,19 +1375,52 @@ func (k *kubernetesClient) configureStorage(
...
+
+	func() {
+		for i, v := range podSpec.InitContainers {
+			if v.Name == initContainer.Name {
+				podSpec.InitContainers[i] = initContainer
+				return
+			}
+		}
+		podSpec.InitContainers = append(podSpec.InitContainers, initContainer)
+	}()
```
[the remark this answers is not in the document]
Comment (hpidcock) [reply caas/kubernetes/provider/k8s.go:1392]: I'd just pull this out to a replaceOrUpdateContainer function.

[the remark this answers is not in the document]
Comment (hpidcock) [reply caas/kubernetes/provider/k8s.go:1392]: Also this is one of those strange cases goto works well for.. as much as I hate goto.
```
	for i, v := range podSpec.InitContainers {
		if v.Name == initContainer.Name {
			podSpec.InitContainers[i] = initContainer
			goto added
		}
	}
	podSpec.InitContainers = append(podSpec.InitContainers, initContainer)
added:
```

[the remark this answers is not in the document]
Comment (hpidcock) [reply caas/kubernetes/provider/k8s.go:1392]: Or when you can determine its the last iteration.
```
for i, v := range podSpec.InitContainers {
	if v.Name == initContainer.Name {
		podSpec.InitContainers[i] = initContainer
		break
	}
	if i+1 == len(podSpec.InitContainers) {
		podSpec.InitContainers = append(podSpec.InitContainers, initContainer)
	}
}
```

[the remark this answers is not in the document]
Comment (hpidcock) [reply caas/kubernetes/provider/k8s.go:1392]: Or break labels
```
out:
for {
	for i, v := range podSpec.InitContainers {
		if v.Name == initContainer.Name {
			podSpec.InitContainers[i] = initContainer
			break out
		}
	}
	podSpec.InitContainers = append(podSpec.InitContainers, initContainer)
	break
}
```

File: worker/caasunitprovisioner/application_worker.go (hunk 2 above)
Comment (hpidcock) [reply worker/caasunitprovisioner/application_worker.go:102]: Opertator must be a kind of root vegetable. :potato:

File: caas/kubernetes/provider/k8s.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ycliuhw) [reply caas/kubernetes/provider/k8s.go]: I don't like `goto` as well. It's bad for readability.

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**